### PR TITLE
[EXTENDED] **11** Parallelized multiple right-hand side support, fully typed `tools`

### PR DESCRIPTION
**Pull request by MothNik**
**⚠️ This pull request can fully replace #26 which can basically be closed ⚠️**

## Changes

This pull request is an improved version of #26, so please refer to the basic changes in there. On top of these changes, this pull request

- fully types the `tools`-module
- parallelizes the Cython-implementation which reqiured
     - restricting the Cython level interface in `solver.pyxd` so that the solvers only accept C-contiguous Arrays. This will **not be backwards compatible** because it could already be that users pass Fortran-contiguous Arrays (maybe without even knowing).
     - refactoring the C-level functions to deal with pointers rather than memoryviews which allows to release the GIL and thus parallelize the right-hand side problems in the first place
    - changing the `setup.py` to include OpenMP
- adds the `worker`-argument to the Python interface

Since all that is a breaking change and the Cython interface is not backwards-compatible, I suggest that this is a **major version jump from 1 to 2 (maybe as an $\alpha$-version of it?)**.
Again, I tried to update the changelog, but it might be that you need to still look over it.

## Tests

The installation and parallelisation were tested on Windows 11 and Ubuntu (WSL). ✅
Tests now cover serial and parallelel solves, but they have to run with
```bash
pytest --cov=pentapy .\tests --cov-report html -x 
```
now to prevent interfering with the multithreading. ✅ 
Depending on the OS, it might be that the doctests of `tests/util_funcs` fail because the Array output includes the `dtype` on Windows while it does not on Linux. ❌
Given that the C-implementation is already very fast in serial, the parallelization does not cause a quantum leap. With 8 threads on a relatively old laptop (there it is, the grain of salt 🧂), I observe a threefold speedup for huge systems (1,000 x 1,000 with 10,000 right-hand sides):

image

On massively parallel systems, this will give better speedups.
However, it does not negatively affect the serial solves when `workers=1`:

ParallelizedSingleRun

### CODE REVIEW

**Review by MothNik — COMMENTED** [review]
Guys, you really declared war on making code maintainable 😅 How should anybody infer that `np.array` is used for both converting lists to Arrays and making a copy at the same time?
I now added a comment and made the tests cover this.

**Comment by MothNik** [comment]
The [last commit](https://github.com/GeoStat-Framework/pentapy/pull/27/commits/aef7d0f23694e9c37895df185b2aebd01d7ef089) currently does not change the internal logic at all.
However, it already forms the foundation of future validations, e.g., by means of the condition number of the matrix. 
Right now, `pentapy` does never assess or allow the user to assess the quality of the solves because the result not being all `np.nan`, this does not mean that the solve was meaningful.
If the user wants to have the results validated, one could introduce `validate` to the Python high-level interface of `solve`.

File: src/pentapy/core.py [hunk 1]
```diff
@@ -169,29 +169,33 @@ def solve(
...
 
-            # if there was only a 1D right-hand side, the result has to be flattened
-            sol, info = solver_func(  # NOTE: info is for potential future validation
-                np.ascontiguousarray(mat_flat),
-                np.ascontiguousarray(rhs),
-                workers,
-                False,  # NOTE: this can enable validation in the future
-            )
+        # ... and the solver is called
+        sol, info = solver_func(
```
Comment (MothNik) [inline src/pentapy/core.py:180]: I removed the `validate` and demonstrated how `info` can be leveraged instead. This allowed to remove the zero-division problem described in #23 without adding another flag.
Everything is now handled on Cython-level.

**Comment by MothNik** [comment]
To be able to cope with #28 and #29 in the future without compromising anything on the solvers side, I rearranged the Cython core (not the interface) as follows:
- the `c_penta_solver1` and `c_penta_solver2` interface are now calling two common functions that
    - factorize the matrix `A` with the respective PTRANS-factorization (I for the first, II for the second)
    - run the (parallelized) loop to solve the transformed right-hand side with this factorization, again with the respective solver
 
   This allowed the removal of the redundant logic (factorize -> solve) that was implemented in both the `c_penta_solver1` and `c_penta_solver2` interface, thereby removing one critical potential point of failure.
- these two common functions are now responsible for all memory allocation and they return the freshly allocated memoryviews of the factorized matrix and the result, respectively. Before that, memory allocation was handled via `np.empty` in  both the `c_penta_solver1` and `c_penta_solver2` interface.
    This in turn now comes with the added bonus, that other interface functions for 
    - factorization
    - solving a right-hand side given an already factorized matrix
    
    can easily be implemented on top and just accept the fresh memoryviews that are return by the common functions. All that's required now for such an interface function is that it calls the common function and casts the returned memoryview to an Array. 
    With these changes, the computation 
    - of the determinant (#28 ),
    - the inverse's main diagonal (#29 ),
    - or just a solve of another right-hand side
  
     can be triggered from Python and Cython after a factorization was obtained. 
   
This is probably a bit too abstract (I barely know how phrase it), so please let me provide the following  
**Analogy Example**
In the basic principle, this is identical to how the banded Cholesky decomposition of SciPy works. For solving a system, there is [`scipy.linalg.solveh_banded`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.solveh_banded.html). However, there is also [`scipy.linalg.cholesky_banded`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.cholesky_banded.html) that can perform a banded Cholesky factorization and [`scipy.linalg.cho_solve_banded`](https://docs.scipy.org/doc/scipy/reference/generated/scipy.linalg.cho_solve_banded.html#scipy.linalg.cho_solve_banded) that can take the banded Cholesky factorization to solve one or more systems.

As of now, this PR basically implements the `scipy.linalg.solveh_banded`-analaogy of `pentay` (factorize and solve in one go), but on the Cython-side I arranged everything that it's super simple to also add the analogies of `scipy.linalg.cholesky_banded` (factorize only) and `scipy.linalg.cho_solve_banded` (solve system using a given factorization). But this would probably explode the frame of this already big PR 🤯

**Comment by MothNik** [comment]
@MuellerSeb Sorry for all these iterations. It tooks me some time to get the right strategy for Cython, the GIL, Memory Allocation, the Parallelization, and a testing framework (they all pass still ✅).
But now I feel that this PR is ready for review and I would highly appreciate your feedback 😄 I really wanted to keep everything flexible for future adjustments. It might be true that `pentapy` is "only" there for solving pentadiagonal systems, so the number of features will never be as mindblowing as SciPy, but playing with the factorizations (like with the determinant) is a common step in mathematical problems and I really wanted to pave way for that 🛣️

File: CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,49 @@
...
 
+## [2.0.0] - 2024-06
+
+See [#27](https://github.com/GeoStat-Framework/pentapy/pull/27)
+
+### Breaking Changes
+
+- fully parallelized the Cython implementation of PTRANS-I and PTRANS-II for single and multiple right-hand sides support that can now be enabled via the new ``workers`` parameter in ``pentapy.solve`` (default: 1)
+- fully typed the ``pentapy.tools`` module
+- updated the **Cython low level interfaces** to PTRANS-I and PTRANS-II to **only accept C-contiguous arrays** (not backwards compatible)
```
Comment (MuellerSeb) [inline CHANGELOG.md:13]: This is the only reason for the major version jump, right? Do you want an intermediate 1.4 release? If not, I would vote for merging the changes of 1.4 and 2.0 here.

File: README.md [hunk 3]
```diff
@@ -107,7 +107,8 @@ Have a look at the script: [``examples/03_perform_simple.py``](https://github.co
 
 ## Requirements:
 
-- [NumPy >= 1.14.5](https://www.numpy.org)
+- [NumPy >= 1.20.0](https://www.numpy.org)
+- [psutil >= 5.8.0](https://psutil.readthedocs.io/en/latest/) (for parallelisation)
```
Comment (MuellerSeb) [inline README.md:111]: `/en/latest/` needed?

File: docs/source/index.rst [hunk 4]
```diff
@@ -94,7 +94,8 @@ The performance plot was created with ``perfplot`` (`link <https://github.com/ns
 Requirements
 ============
 
-- `Numpy >= 1.14.5 <http://www.numpy.org>`_
+- `Numpy >= 1.20.0 <http://www.numpy.org>`_
+- `psutil >= 5.8.0 <https://psutil.readthedocs.io/en/latest/>`_
```
Comment (MuellerSeb) [inline docs/source/index.rst:98]: `/en/latest/` needed?

File: pytest.ini [hunk 5]
```diff
@@ -0,0 +1,2 @@
+[pytest]
+addopts = --doctest-modules
```
Comment (MuellerSeb) [inline pytest.ini:2]: Can't this file content be part of pyproject.toml?

File: pyproject.toml [hunk 6]
```diff
@@ -35,32 +35,10 @@ classifiers = [
...
-test = ["pytest-cov>=3"]
-check = [
-  "black>=24,<25",
-  "isort[colors]",
-  "pylint",
-  "cython-lint",
-]
+[tool.setuptools.dynamic]
+dependencies = {file = ["requirements/base.txt"]}
+optional-dependencies = {scipy = {file = ["requirements/scipy.txt"]}, umfpack = {file = ["requirements/umfpack.txt"]}, all = {file = ["requirements/all.txt"]}, doc = {file = ["requirements/doc.txt"]}, test = {file = ["requirements/test.txt"]}, check = {file = ["requirements/check.txt"]}}
```
Comment (MuellerSeb) [inline pyproject.toml:42]: I have to admit, I am not a fan of these req files. I would like to use my veto here to have the deps in the pyproject.toml file.

File: setup.py [hunk 7]
```diff
@@ -1,23 +1,34 @@
...
 import numpy as np
 from Cython.Build import cythonize
 from setuptools import Extension, setup
 
+if sys.platform.startswith("win"):
+    openmp_arg = "/openmp"
+else:
+    openmp_arg = "-fopenmp"
+
+Cython.Compiler.Options.annotate = True
```
Comment (MuellerSeb) [inline setup.py:16]: Please use https://github.com/astropy/extension-helpers/ here an only activate it with a env-var. OpenMP shouldn't be used by default. See here for example: https://discuss.python.org/t/conflicting-binary-extensions-in-different-packages/25332

See here for reference: https://github.com/GeoStat-Framework/GSTools/blob/v1.6.0/setup.py

File: src/pentapy/__init__.py [hunk 8]
```diff
@@ -44,7 +44,7 @@
 
 try:
     from pentapy._version import __version__
-except ImportError:  # pragma: nocover
+except ImportError:  # pragma: no cover
```
Comment (MuellerSeb) [inline src/pentapy/__init__.py:47]: good catch

File: src/pentapy/errors.py [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+Auxiliary errors for the pentapy package.
+
+"""
+
+# === Imports ===
+
+from enum import Enum
+
+
+class PentaPyErrorMessages(str, Enum):
```
Comment (MuellerSeb) [inline src/pentapy/errors.py:11]: Cool!

Comment (MothNik) [reply src/pentapy/errors.py:11]: I really enjoy working with `Enum`s. My colleagues showed me and since then I'm hooked because it's such a straightforward way of making the codebase more maintainable, safe, and structured 🤩

File: pyproject.toml (hunk 6 above)
Comment (MothNik) [reply pyproject.toml]: Alright, that is just a preference thing. I don't have a super strong opinion about this.
Reverted with [this commit](https://github.com/GeoStat-Framework/pentapy/pull/27/commits/d1433f949775a57ef3cf4a426f32dd2cd99d76a0) 👍

File: pytest.ini (hunk 5 above)
Comment (MothNik) [reply pytest.ini]: I didn't find it everytime I was looking and now it popped up right away.
Moved it to the `pyproject.toml` in [this commit](https://github.com/GeoStat-Framework/pentapy/pull/27/commits/d1433f949775a57ef3cf4a426f32dd2cd99d76a0) 👍

File: CHANGELOG.md (hunk 2 above)
Comment (MothNik) [reply CHANGELOG.md:12]: I would say so given that the C-interface is now way more restrictive.
However, I just closed #26 which would be version 1.4 and merged the two changelogs in [this commit](https://github.com/GeoStat-Framework/pentapy/pull/27/commits/edb4f337defb7437ace62d8c8d15c699a061d758).
So, only a v2.0 would be needed to be released 👍

File: README.md (hunk 3 above)
Comment (MothNik) [reply README.md]: I completely dropped `psutil` with [this commit](https://github.com/GeoStat-Framework/pentapy/pull/27/commits/edb4f337defb7437ace62d8c8d15c699a061d758) in favour of dynamic evaluation with OpenMP :+1:

File: docs/source/index.rst (hunk 4 above)
Comment (MothNik) [reply docs/source/index.rst]: I completely dropped `psutil` with [this commit](https://github.com/GeoStat-Framework/pentapy/pull/27/commits/edb4f337defb7437ace62d8c8d15c699a061d758) in favour of dynamic evaluation with OpenMP :+1:

File: setup.py (hunk 7 above)
Comment (MothNik) [reply setup.py]: Alright, I tried to follow your workflow there and I got it to work in [this commit](https://github.com/GeoStat-Framework/pentapy/pull/27/commits/edb4f337defb7437ace62d8c8d15c699a061d758) 👍 

I
- changed the `workers` to `num_threads` as you said it's debatable and I had no strong opinion on that
- transitioned the `num_threads`-evaluation to the Cython level to drop `psutil` completely. I went slightly different here to also handle negative numbers for automatic evaluation.
- adapted all the variables and tests to the new `num_threads`
- made the OpenMP-linking based on the environment variable (tested all settings (disabled implicitly, disabled explicitly, enable) and the multithreading does not give/give the speedup according to the setting ✅)

File: src/pentapy/errors.py (hunk 9 above)
Comment (MuellerSeb) [reply src/pentapy/errors.py:11]: That is true. Also started using it more with a new project. Especially `IntEnum` can be so handy.

File: setup.py (hunk 7 above)
Comment (MothNik) [reply setup.py]: @MuellerSeb Is there anything that still needs to be modified for the CI workflow for publishing like this one here?
```yml
      - name: Install GSTools
        env:
          GSTOOLS_BUILD_PARALLEL: 1
```

Sorry, I never worked with real Python packaging to make it accessible as Open Source package 😬

**Comment by MuellerSeb** [comment]
@MothNik sorry for a late reply here. I have to admit that this is awesome but totally overwhelming. The last year was packed with private events and the whole stack of pentapy->anaflow->welltestpy really turned into a lowlevel side project where I have next to no time for feature development.

At this point I got the feeling, that I am/was blocking your work and I feel sorry to not support your enthusiasm extending this little package.

Since it is almost a year with no progress, I guess you already moved on and lost interest in pushing this forward, right?

If not, I am a bit puzzled how to move forward, since my time schedule wont change anytime soon. Maybe we could have a short chat at some point.

**Comment by MothNik** [comment]
@MuellerSeb no worries! You communicated back then that you would not have a lot of time.
I also didn't have a lot of time, and I think it's totally fine. Just take your time, and if you are still interested in the features, I'm here to assist.

In the meantime, I indeed moved on and released [`pybandee`](https://pypi.org/project/pybandee/), which is, for now, basically a `numba`-compatible version of `pentapy` without a proper high-level interface. It takes basically the same approach as `pentapy`, but uses a C-oriented matrix layout and also offers functions for computing the log-determinant and the diagonals of the inverse matrix (I opened issues for this here as well). However, I will also not have time to support it properly in the near future.

Long story short: don't beat yourself up! I'm here if you need assistance.

**Comment by MuellerSeb** [comment]
Hey great! So maybe we could add pybandee as a new backend in the future?
